cpp/src/groupby/hash/shared_memory_aggregator.cuh

```diff
@@ -0,0 +1,263 @@
+                             cudf::column_device_view source,
+                             cudf::size_type source_index) const noexcept
+  {
+    using Target          = cudf::detail::target_type_t<Source, cudf::aggregation::SUM_OF_SQUARES>;
+    Target* target_casted = reinterpret_cast<Target*>(target);
+    auto value            = static_cast<Target>(source.element<Source>(source_index));
+    cudf::detail::atomic_add(&target_casted[target_index], value * value);
+
+    if (target_null[target_index]) { target_null[target_index] = false; }
+  }
+};
+
+template <typename Source>
+struct update_target_element_shmem<
+  Source,
+  cudf::aggregation::PRODUCT,
+  cuda::std::enable_if_t<cudf::detail::is_product_supported<Source>()>> {
+  __device__ void operator()(cuda::std::byte* target,
+                             cudf::size_type target_index,
+                             bool* target_null,
+                             cudf::column_device_view source,
+                             cudf::size_type source_index) const noexcept
+  {
+    using Target          = cudf::detail::target_type_t<Source, cudf::aggregation::PRODUCT>;
+    Target* target_casted = reinterpret_cast<Target*>(target);
+    cudf::detail::atomic_mul(&target_casted[target_index],
+                             static_cast<Target>(source.element<Source>(source_index)));
+
+    if (target_null[target_index]) { target_null[target_index] = false; }
+  }
+};
+
+template <typename Source>
+struct update_target_element_shmem<
+  Source,
+  cudf::aggregation::COUNT_VALID,
+  cuda::std::enable_if_t<
+    cudf::detail::is_valid_aggregation<Source, cudf::aggregation::COUNT_VALID>()>> {
+  __device__ void operator()(cuda::std::byte* target,
+                             cudf::size_type target_index,
+                             bool* target_null,
+                             cudf::column_device_view source,
+                             cudf::size_type source_index) const noexcept
+  {
+    using Target          = cudf::detail::target_type_t<Source, cudf::aggregation::COUNT_VALID>;
+    Target* target_casted = reinterpret_cast<Target*>(target);
+    cudf::detail::atomic_add(&target_casted[target_index], Target{1});
+  }
+};
+
+template <typename Source>
+struct update_target_element_shmem<
+  Source,
+  cudf::aggregation::COUNT_ALL,
+  cuda::std::enable_if_t<
+    cudf::detail::is_valid_aggregation<Source, cudf::aggregation::COUNT_ALL>()>> {
+  __device__ void operator()(cuda::std::byte* target,
+                             cudf::size_type target_index,
+                             bool* target_null,
+                             cudf::column_device_view source,
+                             cudf::size_type source_index) const noexcept
+  {
+    using Target          = cudf::detail::target_type_t<Source, cudf::aggregation::COUNT_ALL>;
+    Target* target_casted = reinterpret_cast<Target*>(target);
+    cudf::detail::atomic_add(&target_casted[target_index], Target{1});
```

Comment:
It seems this will only count valid rows since the enable-if requires `is_valid_aggregation`?

Reply:
I think I see what you mean. `is_valid_aggregation` is a compile-time trait that verifies if a specific aggregation type can be applied to the given column data type. It doesn't check the nullability of the input data.

Reply:
Ok. I'm puzzled why `COUNT_ALL` and `COUNT_VALID` appear to have the same implementation.
Both just add 1 to the target.
Shouldn't the `COUNT_VALID` only add 1 if the `source.valid(source_index)`?

Reply:
Right, I've updated the code to check the nullability before explicit instantiations. https://github.com/PointKernel/cudf/blob/b21bc12b8d832fc91311a21b48c72cc6fd68bf54/cpp/src/groupby/hash/shared_memory_aggregator.cuh#L256-L258. The goal was to reduce duplication, but it seems to have impacted clarity. I can revert the change if you think it's not worth it.

Reply:
Ok, I saw that but did not make the connection. 
I think just a simple comment here saying the null is checked before this is called and maybe referenced where that occurs as well.